arangod/Pregel/GraphStore.cpp

```diff
@@ -156,63 +181,38 @@ void GraphStore<V, E>::loadShards(WorkerConfig* config,
   rest::Scheduler* scheduler = SchedulerFeature::SCHEDULER;
   scheduler->post([this, scheduler, callback] {
     uint64_t vertexOffset = 0;
-    std::unordered_map<ShardID, uint64_t> shardSizes = _preallocateMemory();
     // Contains the shards located on this db server in the right order
-    std::map<CollectionID, std::vector<ShardID>> const& vertexCollMap =
-    _config->vertexCollectionShards();
-    std::map<CollectionID, std::vector<ShardID>> const& edgeCollMap =
-    _config->edgeCollectionShards();
-    
-    // calculating sum of all ith edge shards and set it as
-    // starting offset for all i+1 edge shards
-    for (auto const& pair : edgeCollMap) {
-      std::vector<ShardID> const& edgeShards = pair.second;
-      if (_edgeShardsOffset.size() == 0) {
-        _edgeShardsOffset.resize(edgeShards.size()+1);
-        std::fill(_edgeShardsOffset.begin(), _edgeShardsOffset.end(), 0);
-      } else {
-        TRI_ASSERT(_edgeShardsOffset.size() == edgeShards.size()+1);
-      }
-      for (size_t i = 0; i < edgeShards.size(); i++) {
-        _edgeShardsOffset[i+1] += shardSizes[edgeShards[i]];
-      }
-    }
+    // assuming edges are sharded after _from, vertices after _key
+    // then every ith vertex shard has the corresponding edges in
+    // the ith edge shard
     
-    for (auto const& pair : vertexCollMap) {
-      std::vector<ShardID> const& vertexShards = pair.second;
-      for (size_t i = 0; i < vertexShards.size(); i++) {
+    auto shards = _allocateSpace();
+    for (auto& pair : shards) {
+      for (VertexShardInfo& info : pair.second) {
         // we might have already loaded these shards
-        if (_loadedShards.find(vertexShards[i]) != _loadedShards.end()) {
+        if (_loadedShards.find(info.vertexShard) != _loadedShards.end()) {
           continue;
         }
-        
-        ShardID const& vertexShard = vertexShards[i];
-        std::vector<ShardID> edgeLookups;
-        // distributeshardslike should cause the edges for a vertex to be
-        // in the same shard index. x in vertexShard2 => E(x) in edgeShard2
-        for (auto const& pair2 : edgeCollMap) {
-          std::vector<ShardID> const& edgeShards = pair2.second;
-          if (vertexShards.size() != edgeShards.size()) {
-            THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_BAD_PARAMETER,
-                                           "Collections need to have the same number of shards");
-          }
-          edgeLookups.push_back(edgeShards[i]);
-        }
-        
-        _loadedShards.insert(vertexShard);
+        _loadedShards.insert(info.vertexShard);
         _runningThreads++;
-        scheduler->post([this, i, vertexShard, edgeLookups, vertexOffset] {
+        TRI_ASSERT(info.numVertices > 0);
+        TRI_ASSERT(vertexOffset < _index.size());
+        TRI_ASSERT(info.edgeDataOffset < _edges->size());
+        scheduler->post([this, &info, vertexOffset] {
```

Comment:
Theoretically `info` may become invalid if the callback is executed by the Scheduler and the surrounding function `loadShards` that defines `info` is left. This is unlikely though (may happen if `loadShards` throws an exception after posting jobs). Can happen if the `_loadedShards.insert(...)` calls fail with OOM or such.
It would be safer to *copy* info into the callback.

Reply:
nope, there is  `while (_runningThreads > 0) { std::this_thread::sleep_for(std::chrono::microseconds(5000));}` belowe

Reply:
Yes, but if the mentioned insert throws, then the while loop will not be reached at all and the contents of `info` will be undefined.